### ISSUE TITLE
Empty container displayed when grouping notifications

### ISSUE THREAD

**Issue by tad3j** [notifee/react-native-notifee#72]
EDIT: After looking further into this in my second post I realize I may be misusing notifee (by caliing group container creation twice) , so I guess this is more like a future request.

I'm grouping my notifications by postId, and I'm noticing that in some cases notifee is displaying an empty group container, without any notification inside it (see image). After notification (grouped) is pressed and I re-open the notifications tray...I see an empty container that was used for grouping notifications (in place of 1st notification they are still displayed as grouped, but in place of 2nd notification there is an empty group container). Also if I close all of the notifications inside group, that group container will still there empty, so I have to swipe close that one as well.

I don't think this is expected behavior? Or do we need to handle group containers on our own?

![image](https://user-images.githubusercontent.com/4150904/83383862-44c83b00-a3e6-11ea-8202-e4e8cc32aa3c.png)

Here is my code for displaying notification (part of `onMessageReceived`):
```
//GROUPING NOTIFICATIONS: https://notifee.app/react-native/docs/android/grouping-and-sorting
  const summaryForChannelGroup = {
    channelId,
    groupId: notificationPayload.data.postId, //group by post id
    groupAlertBehavior: AndroidGroupAlertBehavior.SUMMARY,
    ...commonDeliverOptions,
    lights: [AndroidColor.RED, 300, 600],
  }

  //SUMMARY NOTIFICATION: not actually displayed until we have more than 1 notification,
  // then it acts like a container for group of notifications (grouped by groupId)
  await notifee.displayNotification({
    subtitle: notificationPayload.channelName,
    android: {
      groupSummary: true,
      ...summaryForChannelGroup,
    },
  })

  //ACTUAL NOTIFICATION
  notificationPayload.android = {
    ...summaryForChannelGroup,
    actions: [
      {
        title: 'Dismiss',
        pressAction: {
          id: 'dismiss',
        },
      },
    ],
  }
  await notifee.displayNotification(notificationPayload)
```

**Comment by khlling** [comment]
'''
Also if I close all of the notifications inside group, that group container will still there empty, so I have to swipe close that one as well.
'''

Similar behaviour to this?
https://github.com/notifee/react-native-notifee/issues/62

I've found that if you just call _displayNotification_ with data where:
data.android.groupSummary: true

Then it creates an empty group notification. I'm assuming that we have to handle the creation of the group manually i.e. create the group first then subsequently any notifications with the same groupId will be grouped together.

This my interpretation of the documentation:

```
Grouped notifications share a unique identifier, which can be provided to the groupId property. Before displaying notifications in a group, you must first create a "summary" notification which holds all future notifications with the same ID.
```

source: https://notifee.app/react-native/docs/android/grouping-and-sorting

**Comment by khlling** [comment]
I'm not an expert as I'm currently exploring the source code and the module myself, but from my understanding what you've stated here is correct:

> Hm...I guess we should do the counting of unread notifications on our side and avoid calling group creation if we have at least one unread notification? I've simulated that by manually removing code that creates group container (before 2nd notification) and this time all looks fine.

And I completely agree that it would helpful if the module kept the state of the notifications  and sorted this all out for you for completeness. Having said that I would be surprised if this wasn't already in the roadmap.

You have to remember that Notifee is still in alpha the team are still gathering feedback from users. Having said that I echo your point and I'm going to have an experiment/dig around in the source code when I have spare time to see what changes could be made.

**Comment by tad3j** [comment]
> 
> 
> I'm not an expert as I'm currently exploring the source code and the module myself, but from my poking around what you've stated here is correct:
> 
> > Hm...I guess we should do the counting of unread notifications on our side and avoid calling group creation if we have at least one unread notification? I've simulated that by manually removing code that creates group container (before 2nd notification) and this time all looks fine.

Me neither, my first RN project. :) ...I've bought notifee like 3 months ago, so I though it may be time to report issue that I had from start (didn't notice one that's kind of related exist).

> And I completely agree that it would helpful if the module kept the state of the notifications and sorted this all out for you for completeness. Having said that I would be surprised if this wasn't already in the roadmap.
> 
> You have to remember that Notifee is still in alpha the team are still gathering feedback from users. Having said that I echo your point and I'm going to have an experiment/dig around in the source code to see what if the current implementation could support that.

Right, I would think so as well. I'm aware, I hope its get improved soon.

**Comment by Ehesp** [comment]
Hey, thanks for the discussion / investigation.

Android will automatically group notifications after some time (depending on some device details), however the general idea is that you shouldn't be creating an empty container. 

For example, take the following code:

```js
    await notifee.displayNotification({
      id: 'email-summary',
      title: 'Emails',
      subtitle: '2 Unread Emails',
      android: {
        channelId: 'default',
        groupSummary: true,
        groupId: 'email',
      },
    });

    await notifee.displayNotification({
      title: 'New Email',
      body: 'Some email content',
      subtitle: 'Unread',
      android: {
        channelId: 'default',
        groupId: 'email',
      },
    });

    await notifee.displayNotification({
      title: 'New Email',
      body: 'Some email content',
      subtitle: 'Unread',
      android: {
        channelId: 'default',
        groupId: 'email',
      },
    });
```

If you run this, it'll create 2 notifications grouped inside one. You are able to wipe away the entire group, or individual ones, however if only one remains it also removes the group notification (same with pressing them).

> When I dismiss last notification with swiping, notification disappears, but both empty containers get shown again.

Could you try adding a custom ID to the group-summary notification (like shown above)? I can't replicate that behaviour.

**Comment by khlling** [comment]
I tried the snippet you've proved and your example works fine as and as you describe. Including dismissing the entire group and removing the group with the last notification.

However, something I'd like clarification on is around the implementation of grouping client-side (on a device) vs server-side.

May be I'm misunderstanding this but the group summary notification (i.e. where groupSummary: true), needs to be created before the corresponding notifications can appear within the group notification. Is it your intention that grouping is meant to be handled server-side instead?

For example if I want to send 2 FCM then I first have to send the summary notification then the follow up 2 notifications and not create the group summary first as per the creation of channels.

**Comment by tad3j** [comment]
Thanks @Ehesp for your reply. I'm using the same channel for both notifications, it's set with `channelId: ``${groupId}``,` (sorry, didn't share that part).

You weren't able to reproduce the issue since you are using notifee in a way that most likely won't get used in a real world app, where notifications arrive in random times. So for real implementation one would use only 2 calls to `notifee.displayNotification`, 1 to create the container and 1 to put notification inside that container and after that, notification handler ends. The problem appears when next notification arrives and user hasn't closed the previous one....in this case I am again making 2 calls as above. 1st call to `notifee.displayNotification` (with `groupSummary: true,`) will then create a new empty container (groupId and channelId passed to notifee are the same for both notifications).

That's why as I mention above the only workaround I see ATM is to count unread notification and only call `notifee.displayNotification` with `groupSummary: true,` if that counter is at 0 (and has to be done for each container). That's why I think it would be much better if notifee just wouldn't display an empty container if one already exists for the same `groupId`.

**Comment by Ehesp** [comment]
Thanks for the explanations!

So I think this is where an `id` can be used.

TLDR; Always create a summary notification with the same ID every time a notification is delivered.

1. Server sends an FCM to the device:

```js
// FCM handler

// Create a group summary holder
await Notifee.displayNotification({
  id: "group", // important
  title: "Group Chat",
  android: {
    channelId: "default",
    groupSummary: true,
    groupId: "123"
  }
});

// This notification would be created from the FCM payload: 
await Notifee.displayNotification({
  title: "Message 1",
  android: {
    channelId: "default",
    groupSummary: true,
    groupId: "123"
  }
});
```

2. The user swipes away the single message / group on the device. No notifications exist.

3. Server sends another FCM to the device:

```js
// FCM handler

// Create a group summary holder
await Notifee.displayNotification({
  id: "group", // important
  title: "Group Chat",
  android: {
    channelId: "default",
    groupSummary: true,
    groupId: "123"
  }
});

// This notification would be created from the FCM payload: 
await Notifee.displayNotification({
  title: "Message 2",
  android: {
    channelId: "default",
    groupSummary: true,
    groupId: "123"
  }
});
```

4. Server sends another FCM, however this time the user hasn't swiped the existing notification/group away:

```js
// FCM handler

// This won't actually trigger a new group because the `id` field is the same.
// We could however, update the text of the group summary
await Notifee.displayNotification({
  id: "group", // important
  title: "Group Chat",
  android: {
    channelId: "default",
    groupSummary: true,
    groupId: "123"
  }
});

// This notification would be created from the FCM payload: 
await Notifee.displayNotification({
  title: "Message 3",
  android: {
    channelId: "default",
    groupSummary: true,
    groupId: "123"
  }
});
```

5. The device would still show 1 group summary container (locked by the `id`), however hold Messages 2 & 3 as expected.

**Comment by khlling** [comment]
This works!

```
await Notifee.displayNotification({
  id: "group", // important
  title: "Group Chat",
  android: {
    channelId: "default",
    groupSummary: true,
    groupId: "123"
  }
});

// This notification would be created from the FCM payload: 
await Notifee.displayNotification({
  title: "Message 1",
  android: {
    channelId: "default",
    groupId: "123"
  }
});
```
I think you made a typo by copying in groupSummary: true for the actual notification.

So in summary you can send the group message multiple times as long as the groupId's match and notifee will handle the grouping :)

**Comment by Ehesp** [comment]
Good catch, just edited my examples.

Yeah, as long as you send the same custom ID it'll update what's already there without creating something new. By default, a random ID is generated which in this case you don't want :)

**Comment by yotam-nots** [comment]
@Ehesp 
can you please comment on the somewhat related issue https://github.com/notifee/react-native-notifee/issues/62
Basically, when canceling grouped notifications using `notifee.cancelNotification`, the group doesn't get remove with the last notification cancelation.

**Comment by tad3j** [comment]
@Ehesp, thank you...this is amazing. Didn't think of using id for container. After adding it issues I described are gone.

I'm still wondering, what's the difference then between `id` and `groupId`?  

From docs: 
GroupId
`Grouped notifications share a unique identifier, which can be provided to the groupId property.`
- tells how to group notifications? But aren't all notifications passed to container with same ID grouped anyway?
ID:
`IDs can later be used to update any notification which is still present on the user's device`).
- tells which notification container to update?

**Comment by Ehesp** [comment]
A `groupId` links the notifications together so the device can group them up in the notification shade. The notification `id` is per-notification, allowing you to update an existing ones on the device.

I'll close this one since it seems to be solved.

**Comment by tad3j** [comment]
Even with that in place, I'm still seeing empty container on production build, on real device. 

To reproduce, send a notification, when it arrives click `Dismiss`, and empty container is left after notifee clears notification.

**Comment by helenaford** [comment]
Closing this in favour of #62
